src/hotspot/cpu/s390/stubGenerator_s390.cpp

```diff
@@ -2857,6 +2858,44 @@ class StubGenerator: public StubCodeGenerator {
     return start;
   }
 
+  address generate_nmethod_entry_barrier() {
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "nmethod_entry_barrier");
+
+    address start = __ pc();
+
+    // Save caller's sp, return_pc in frame large enough for vm call
+    __ save_return_pc();
+    __ push_frame_abi160(0);
+
+    // Prep arg for VM call
+    // Create PTR to (R14 - Barrier_Length)
+    // This is the address of the entry point for the compiled fn we are examining.
+    __ z_lay(Z_R1_scratch, -32, Z_R0, Z_R14);           // R1 <- R14 - 32
+    __ z_stg(Z_R1_scratch, _z_abi(carg_2), Z_R0, Z_SP); // SP[abi_carg2] <- R1
+    __ z_la(Z_ARG1, _z_abi(carg_2), Z_R0, Z_SP);        // R2 <- SP + abi_carg2
```

Comment:
Z_ARG1 should point to the address _z_abi16(return_pc) + Z_SP in the caller frame. (Don't generate a copy!) That matches _z_abi16(return_pc) + 160 + Z_SP in the current frame at this point.
In addition, I'm missing save_volatile_gprs & restore_volatile_gprs.

Reply:
> Z_ARG1 should point to the address _z_abi16(return_pc) + Z_SP in the caller frame.

This matches what the PPC implementation does, but when I do the same thing on s390 I get a cache miss in nmethod_stub_entry_barrier (the vm-call). It looked as though CodeCache::find_blob expects the address of the start of the compiled code, so I tried subtracting the size of the barrier from R14 (which currently points to end of the barrier in the compiled frame). After doing this I no longer saw the CodeCache miss.

> I'm missing save_volatile_gprs & restore_volatile_gprs for GP and FP regs.

I had been trying to get the volatile registers saved, but didn't have any luck. I tried it today with your suggestions and it worked like a charm. Not sure what the difference was. Thanks for the pointers.

Reply:
After a bit more investigation, I believe I see the reasoning behind using the suggested computation for the VM-Call's argument. It seems the CodeCache miss is the issue, not the argument, so I am focusing my efforts on understanding what is happening there.

Reply:
You just used the return_pc address of the wrong frame.

Reply:
Right. I see the difference mentioned in your other comment. I have made this change. Thanks for clarifying.